components/scream/src/share/field/field_repository.hpp

```diff
@@ -232,72 +267,118 @@ has_field (const identifier_type& identifier) const {
+  EKAT_REQUIRE_MSG(ptr!=nullptr, "Error! Field not found.\n");
+  return *ptr;
 }
 
 template<typename RealType>
 const typename FieldRepository<RealType>::field_type&
-FieldRepository<RealType>::get_field (const std::string name, const std::string& grid) const {
-  EKAT_REQUIRE_MSG(m_repo_state==RepoState::Closed,"Error! You are not allowed to grab fields from the repo until after the registration phase is completed.\n");
-
-  // Keep track of the number of fields found for this name/grid combo
-  std::vector<identifier_type> f_matches;
-  //  Search field repo for field matching this name.
-  const auto& map = m_fields.find(name);
-  EKAT_REQUIRE_MSG(map!=m_fields.end(), "Error! get_field: " + name + " not found.\n");
-  //  Search subset of field repo for matching gridname.
-  for (const auto& it : map->second) {
-    if ( it.first.get_grid_name()==grid) {f_matches.push_back(it.first);}
-  }
-  // Check to make sure a) the field was found on this grid, and b) only once.
-  EKAT_REQUIRE_MSG(f_matches.size()==1, "Error! get_field: " + name + " found " + std::to_string(f_matches.size()) + " matches on grid " + grid + ".\n");
-  // Use this field id to grab field itself.
-  return get_field(f_matches[0]);
+FieldRepository<RealType>::get_field (const std::string& name, const std::string& grid) const {
+  EKAT_REQUIRE_MSG(m_repo_state==RepoState::Closed,
+      "Error! Cannot get fields from the repo while registration has not yet completed.\n");
+  auto ptr = get_field_ptr(name,grid);
+  EKAT_REQUIRE_MSG(ptr!=nullptr, "Error! Field not found.\n");
+  return *ptr;
 }
 
 template<typename RealType>
-std::set<typename FieldRepository<RealType>::field_type>
+FieldGroup<typename FieldRepository<RealType>::RT>
 FieldRepository<RealType>::
 get_field_group (const std::string& group_name, const std::string& grid_name) const {
-  auto names = m_field_groups.find(group_name);
-  EKAT_REQUIRE_MSG (names!=m_field_groups.end(),
-                    "Error! Group name '" << group_name << "' not found in the repo.\n");
-
-  std::set<field_type> group_fields;
-  for (const auto& fname : names->second) {
-    group_fields.insert(get_field(fname,grid_name));
+  EKAT_REQUIRE_MSG(m_repo_state==RepoState::Closed,
+      "Error! Cannot get field groups from the repo while registration has not yet completed.\n");
+
+  // Create an empty group
+  FieldGroup<RT> group(group_name,grid_name);
+
+  // Allow returning an empty group
+  auto it = m_field_groups.find(group_name);
+  if (it!=m_field_groups.end()) {
+    group.m_info = it->second;;
+    for (const auto& fname : group.m_info->m_fields_names) {
+      auto f = get_field_ptr(fname,grid_name);
+      group.m_fields[fname] = f;
+      // Fetch the bundle fields (if bundled) just once
+      if (group.m_info->m_bundled && group.m_bundle==nullptr) {
+        auto p = f->get_header().get_parent().lock();
+        EKAT_REQUIRE_MSG(p!=nullptr, "Error! Something is amiss with a bundled field group.\n");
```

Comment:
Maybe "a bundled field is missing its parent"?

Reply:
Yeah, it can be phrased better.